src/main/clojure/clara/rules/engine.cljc

```diff
@@ -1748,50 +1748,74 @@
-        ;; of these insertions can be cached and executed as a batch instead of eagerly realizing
-        ;; them.  An external insertion of a fact that matches
-        ;; a negation or accumulator condition can cause logical retractions.
-        (doseq [[alpha-roots fact-group] (get-alphas-fn facts)
-                root alpha-roots]
-          (alpha-activate root fact-group transient-memory transport transient-listener))
-        (external-retract-loop get-alphas-fn transient-memory transport transient-listener))
+    (let [new-pending-operations (conj pending-operations (->PendingUpdate :insertion (if (coll? facts)
+                                                                                        facts
+                                                                                        (into [] facts))))]
 
       (LocalSession. rulebase
-                     (mem/to-persistent! transient-memory)
+                     memory
                      transport
-                     (l/to-persistent! transient-listener)
-                     get-alphas-fn)))
+                     listener
+                     get-alphas-fn
+                     new-pending-operations)))
 
   (retract [session facts]
 
-    (let [transient-memory (mem/to-transient memory)
-          transient-listener (l/to-transient listener)]
-
-      (l/retract-facts! transient-listener facts)
-
-      (binding [*pending-external-retractions* (atom facts)]
-        (external-retract-loop get-alphas-fn transient-memory transport transient-listener))
+    (let [new-pending-operations (conj pending-operations (->PendingUpdate :retraction (if (coll? facts)
+                                                                                         facts
+                                                                                         (into [] facts))))]
 
       (LocalSession. rulebase
-                     (mem/to-persistent! transient-memory)
+                     memory
                      transport
-                     (l/to-persistent! transient-listener)
-                     get-alphas-fn)))
+                     listener
+                     get-alphas-fn
+                     new-pending-operations)))
 
   (fire-rules [session]
 
     (let [transient-memory (mem/to-transient memory)
           transient-listener (l/to-transient listener)]
+
+      ;; We originally performed insertions and retractions immediately after the insert and retract calls,
+      ;; but this had the downside of making a pattern like "Retract facts, insert other facts, and fire the rules"
+      ;; perform at least three transitions between a persistent and transient memory.  Delaying the actual execution
+      ;; of the insertions and retractions until firing the rules allows us to cut this down to a single transition
+      ;; between persistent and transient memory.  There is some cost to the runtime dispatch on operation types here,
```

Comment:
Yeah, it would be almost nonexistent unless you were, say, inserting/retractions tons of facts one at a time which will already perform poorly anyway.  It is basically just a keyword equality check.  That can add up if you do enough of them; we've had cases where doing (identical? some-arg :some-keyword) rather than (= some-arg :some-keyword) in a hotspot sped things up considerably.  I doubt this is a meaningful concern in this particular case though.